libsolidity/experimental/ast/TypeSystem.cpp

```diff
@@ -0,0 +1,357 @@
+		{PrimitiveType::TypeFunction, "tfun", 2},
+		{PrimitiveType::Function, "fun", 2},
+		{PrimitiveType::Itself, "itself", 1},
+		{PrimitiveType::Void, "void", 0},
+		{PrimitiveType::Unit, "unit", 0},
+		{PrimitiveType::Pair, "pair", 2},
+		{PrimitiveType::Sum, "sum", 2},
+		{PrimitiveType::Word, "word", 0},
+		{PrimitiveType::Integer, "integer", 0},
+		{PrimitiveType::Bool, "bool", 0},
+	})
+		m_primitiveTypeConstructors.emplace(type, declareTypeConstructor(name, name, arity, nullptr));
+
+	TypeClass classType = primitiveClass(PrimitiveClass::Type);
+	//TypeClass classKind = primitiveClass(PrimitiveClass::Kind);
+	Sort typeSort{{classType}};
+	m_typeConstructors.at(m_primitiveTypeConstructors.at(PrimitiveType::TypeFunction).m_index).arities = {Arity{std::vector<Sort>{{typeSort},{typeSort}}, classType}};
+	m_typeConstructors.at(m_primitiveTypeConstructors.at(PrimitiveType::Function).m_index).arities = {Arity{std::vector<Sort>{{typeSort, typeSort}}, classType}};
+	m_typeConstructors.at(m_primitiveTypeConstructors.at(PrimitiveType::Itself).m_index).arities = {Arity{std::vector<Sort>{{typeSort, typeSort}}, classType}};
+}
+
+TypeVariable TypeSystem::freshVariable(Sort _sort)
+{
+	size_t index = m_numTypeVariables++;
+	return TypeVariable(index, std::move(_sort));
+}
+
+TypeVariable TypeSystem::freshTypeVariable(Sort _sort)
+{
+	_sort.classes.emplace(primitiveClass(PrimitiveClass::Type));
+	return freshVariable(_sort);
+}
+
+std::vector<TypeEnvironment::UnificationFailure> TypeEnvironment::instantiate(TypeVariable _variable, Type _type)
+{
+	for (auto const& maybeTypeVar: TypeEnvironmentHelpers{*this}.typeVars(_type))
+		if (auto const* typeVar = std::get_if<TypeVariable>(&maybeTypeVar))
+			if (typeVar->index() == _variable.index())
+				return {UnificationFailure{RecursiveUnification{_variable, _type}}};
+
+	Sort typeSort = sort(_type);
+	if (!(_variable.sort() <= typeSort))
+		return {UnificationFailure{SortMismatch{_type, _variable.sort() - typeSort}}};
+
+	auto [_, inserted] = m_typeVariables.emplace(_variable.index(), _type);
+	solAssert(inserted);
+	return {};
+}
+
+experimental::Type TypeEnvironment::resolve(Type _type) const
+{
+	Type result = _type;
+	while (auto const* var = std::get_if<TypeVariable>(&result))
+		if (Type const* resolvedType = util::valueOrNullptr(m_typeVariables, var->index()))
+			result = *resolvedType;
+		else
+			break;
+	return result;
+}
+
+experimental::Type TypeEnvironment::resolveRecursive(Type _type) const
+{
+	return std::visit(util::GenericVisitor{
+		[&](TypeConstant const& _type) -> Type {
+			return TypeConstant{
+				_type.constructor,
+				_type.arguments | ranges::views::transform([&](Type _argType) {
+					return resolveRecursive(_argType);
+				}) | ranges::to<std::vector<Type>>
+			};
+		},
+		[&](TypeVariable const&) -> Type {
+			return _type;
+		},
+		[&](std::monostate) -> Type {
+			return _type;
+		}
+	}, resolve(_type));
+}
```

Comment:
I think I finally figured out why `resolveRecursive()` sometimes does not resolve things when `resolve()` does. It has a bug. For `TypeVariable` it returns the unresolved `_type` from the function scope rather than the resolved variable it gets in the parameter.

Reply:
This is now a part of #14658.